streamer/m3u8_concater.py

```diff
@@ -0,0 +1,955 @@
+              aud_playlists)
+          for aud_playlist in aud_playlists:
+            concat_aud_playlist.content += aud_playlist.content
+            # Add a discontinuity after each period.
+            concat_aud_playlist.content += '#EXT-X-DISCONTINUITY\n\n'
+          # The audio and the stream variant playlist will be exactly the same.
+          concat_var_playlist.target_duration = concat_aud_playlist.target_duration
+          concat_var_playlist.content = concat_aud_playlist.content
+          concat_aud_only_playlists.extend(
+              [concat_aud_playlist, concat_var_playlist])
+    
+    return concat_aud_only_playlists
+  
+  @staticmethod
+  def concat_vid(all_vid_playlists: List[List['MediaPlaylist']],
+                 durations: List[float]) -> List['MediaPlaylist']:
+    """Concatenates multiple video playlists into one multi-period playlist
+    for many resolutions and codecs, it matches the codecs first, then matches
+    the resolutions.
+    
+    It will pick the closest lower resolution for some period(input) if a high
+    enough resolution was not available.
+    
+    A video playlist is a stream variant playlist based on the Packager's output.
+    """
+    
+    # Get all possible video codecs.  We should not use the all codecs available
+    # in the pipeline config, because for some codecs we might have given the
+    # Packager the 'dash_only' flag.
+    codecs: Set[VideoCodec] = set()
+    # Also get all the available resolutions.
+    resolutions: Set[VideoResolution] = set()
+    for vid_playlists in all_vid_playlists:
+      for vid_playlist in vid_playlists:
+        assert isinstance(vid_playlist.codec, VideoCodec)
+        codecs.add(vid_playlist.codec)
+        resolutions.add(vid_playlist.resolution)
+    
+    # Create and initialize a division map.
+    division: Dict[VideoCodec,
+                   Dict[VideoResolution,
+                        List[MediaPlaylist]]] = {}
+    for codec in codecs:
+      division[codec] = {}
+      for resolution in resolutions:
+        division[codec][resolution] = []
+    
+    # In each period do the following:
+    for vid_playlists in all_vid_playlists:
+      # Initialize a mapping between video codecs and a list of resolutions available.
+      codec_division: Dict[VideoCodec, List[MediaPlaylist]] = {}
+      for codec in codecs:
+        codec_division[codec] = []
+      # For every video playlist in this period, append it to the matching codec.
+      for vid_playlist in vid_playlists:
+        assert isinstance(vid_playlist.codec, VideoCodec)
+        codec_division[vid_playlist.codec].append(vid_playlist)
+      for codec in codecs:
+        # Sort the variants from low resolution to high resolution.
+        codec_division[codec].sort(key=lambda pl: pl.resolution)
+        for i, resolution in enumerate(sorted(resolutions)):
+          division[codec][resolution].append(
+              # Append the ith resolution if found, else, append
```

Comment:
So you're not comparing the actual resolutions?  I'm thinking about this abstractly here at first, with something like:

[360p, 720p, 1080p]  +  [720p, 1080p, 4k]

Going by index, you'd end up with something misaligned like:

[360p + 720p, 720p + 1080p, 1080p + 4k]

Instead of something like:

[360p + 720p, 720p + 720p, 1080p + 1080p, 1080p + 4k]

At least in Player, we have to contend with these scenarios.  But I guess here, we don't, since the bottom rungs of the output from Streamer will always align since they came from the same pipeline config, and we only cut off the top rungs based on the input res.  Is that accurate?

Could you explain this in the comment a little?  (If I understood correctly, add to the comment that the lower resolutions will always align because the outputs shared a pipeline config, so this indexing is always safe.)

Reply:
Yes, i am using this perk of the Streamer to make the matching easier, actually i had to do it similar to what you do in the Player with the audio channel matching prior to the 'channel_layout as an input feature' PR.  I was starting with a dictionary full on `None`s for every slot and start filling it appropriately with the channel count it should get for every period.